### DISCUSSION TITLE
Syntactical sugar to reduce repetition in using statements

### DISCUSSION THREAD

**Discussion by epvanhouten** [dotnet/csharplang#1421] [Language Ideas]
Current implementation leads to blocks that are similar to the following:

```
using System;
using System.Net;
using System.Threading;
using System.Threading.Tasks;
```

It would reduce repetition in these types of statement groupings if a syntax similar to the following was possible:

```
using System.{
  this;
  Net;
  Threading.{
    this;
    Tasks;
  };
};
```

**Comment by tannergooding** [comment]
The proposed syntax is almost impossible to read (In my Opinion).

And while you might be able to describe the current syntax as "verbose", it is non-ambiguous and is limited to a few lines at the top of every file (and the IDE adds half these via analyzers/code-fixes, so it's not a lot of typing required either).

**Comment by epvanhouten** [comment]
The above example was intentionally minimal and the styling may not be appropriate. Another example and formatting:

```
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Reflection;
using System.Text;
using System.Threading.Tasks;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Input;
using System.Windows.Media;
using System.Xml;
```

Could be expressed as:

```
using System.{
    this; Collections.Generic; Diagnostics; IO; Reflection; Text; Threading.Tasks; Windows; Xml;
    Windows.{Controls; Input; Media};
  };
```

Which is much more concise than the above.

**Comment by Richiban** [comment]
I like the idea (and will be very familiar to Javascript devs) but I do think it's a bit harder to read.

Maybe if we disallow nesting?
```csharp
using System.{this, Diagnostics, IO, Reflection, Text, Xml};

using System.Collections.Generic;
using System.Threading.Tasks;

using System.Windows.{this, Controls, Input, Media};
```

**Comment by alrz** [comment]
In this space I'd rather we invest on local using directives that actually enable things (https://github.com/dotnet/csharplang/issues/218), e.g. if you have ambiguous extension methods, you can import the desired one in the closer scope.

This on the other hand, makes things less approachable. as @tannergooding mentioned, you don't usually type usings by hand. Also, the simple flat structure is the most copy/paste-friendly syntax, if you need to.

**Comment by bowbahdoe** [comment]
To draw a phrase from the python community "There should be one-- and preferably only one --obvious way to do it."

**Comment by iamcarbon** [comment]
Another possible syntax inspired by Linq.

```
using { Self, Diagnostics, IO, Reflection, Text, Xml } from System;
using { Generic, Specialized } from System.Collections;
using { Tasks } from System.Threading;
using { Models, Views, Controllers } from MyCorp;
```

**Comment by HaloFour** [comment]
@Richiban 

How Scala-esque.

But in Scala the set syntax only works on individual types, not packages.  It's only really useful when you want to import a couple of different classes without having to import the entire namespace via wildcard.  Scala syntax doesn't work on multiple packages that way, and wildcard imports put you into the exact same situation that you have in C#.

If such a syntax were to be considered for C# I think I'd prefer breaking out the parent namespace, though.  I don't like the look of `this`.

```cs
using System;
using System.{Diagnostics, IO, Reflection, Text, Xml};

using System.Collections.Generic;
using System.Threading.Tasks;

using System.Windows;
using System.Windows.{Controls, Input, Media};
```

But then how might aliases work?  Or `static` imports?

**Comment by vbcodec** [comment]
JSON rulez :)

**Comment by VanKrock** [comment]
Who writes usings block manually?
Keep calm and press Alt+Enter or Ctrl + .

**Comment by jnm2** [comment]
> Who writes usings block manually?
> Keep calm and press Alt+Enter or Ctrl + .

This is honestly where I've been with using directives for years.
